### ISSUE TITLE
variable-name-format order

### ISSUE THREAD

**Issue by TheJaredWilcurt** [sasstools/sass-lint#1175]
Sass Lint v1.12.1

```sass
body
    background: $myColor
```

```yml
  variable-name-format:
    - 1
    - convention: camelcase
    - convention-explanation: POTATO!
    - allow-leading-underscore: false
```

when `allow-leading-underscore: false` is above `convention: camelcase` the convention is ignored. I am also unable to get `convention-explanation` to work at all.

**Comment by M1ras** [comment]
I've tested this bug on every release of Sass Lint

- from v1.3.0 (the initial release that supports `variable-name-format`)
- to v1.12.1 (the current release)

and I can confirm every release is affected.

**Comment by M1ras** [comment]
I've fixed the issue and am in the process of opening a separate issue, creating a demonstrative repository, and submitting a pull request.

The issue affects every rule that has subrules since Sass Lint v1.3.0, not just `variable-name-format`. I'm baffled at how it's gone unnoticed for nearly 3 years.

**Comment by adeacetis** [comment]
@M1ras It has been noticed but never formulated this way. It makes way more sense right now. A couple of months ago at work we have noticed the `max-depth` sub rule didn't work which is quite annoying. Will watch this closely.

Thanks for the heads up.

**Comment by DanPurdy** [comment]
Sorry but you've formatted your config incorrectly..

Your example
```yaml
 variable-name-format:
    - 1
    - convention: camelcase
    - convention-explanation: POTATO!
    - allow-leading-underscore: false
```
produces the following json
```json
{
  "variable-name-format": [
    1,
    {
      "convention": "camelcase"
    },
    {
      "convention-explanation": "POTATO!"
    },
    {
      "allow-leading-underscore": false
    }
  ]
}
```

Each rule requires an object containing the rule options, not a single object per option e.g.

```yaml
variable-name-format:
    - 1
    - 
      convention: camelcase
      convention-explanation: POTATO!
      allow-leading-underscore: false
```

gives you the following
```json
{
  "variable-name-format": [
    1,
    {
      "convention": "camelcase",
      "convention-explanation": "POTATO!",
      "allow-leading-underscore": false
    }
  ]
}
```

You can see examples in the documentation for the rules
https://github.com/sasstools/sass-lint/blob/develop/docs/rules/variable-for-property.md#options

And the matching structure in the tests
https://github.com/sasstools/sass-lint/blob/develop/tests/rules/variable-for-property.js#L22

We should probably make this clearer, but it's also worth pointing out that you can just use a json config file now, although currently i don't think some of the upstream consumers i.e. the gulp package accept this. Im really trying to find time to come back and update this project as I have some big improvements for a v2 ready.

If the above doesn't fix your issue let me know.

**Comment by TheJaredWilcurt** [comment]
Cool, but on a related note, YAML is really stupid, why doesn't everyone just use JSON for config files.

**Comment by TheJaredWilcurt** [comment]
I don't think I understand how to tell when something should or should not have a `-` in front of it

```yml
rules:
  brace-style:
    - 2
    - allow-single-line: false
      style: '1tbs'
  mixins-before-declarations:
    - 2
    - exclude:
        - breakpoint
        - mq
  no-qualifying-elements:
   - 2
   - allow-element-with-attribute: true
     allow-element-with-class: false
     allow-element-with-id: false
  no-trailing-whitespace: 1
  property-sort-order:
    - 1
    - order:
        - content
        - position
        - top
  variable-name-format:
    - 1
    - convention: camelcase
      allow-leading-underscore: false
```
is `- breakpoint` `- mq` right, or the `- content` `- position` `- top`? Why do they get their own dashes. Why do some dashes mean they are turned into a number, others an object, and others as what I presume are items in an array?

**Comment by DanPurdy** [comment]
play around with this https://www.json2yaml.com/

I agree YAML isn't the most friendly format but once you understand its layout it's fairly easy. it's less finnicky than JSON imo but i do believe you should be able to use either format. 

The example you posted looks fine.

essentialy the `-` just means a new property and if what follows it is a key value pair then it'll 'put them in an object'
if the `-` precedes a value after a key then it creates elements in an array and similar to `.sass` indentation matters.
